lock-impl/src/test/java/com/palantir/lock/impl/ThreadInfoLockServiceImplTest.java

```diff
@@ -0,0 +1,308 @@
+                            .build());
+            latch.countDown();
+            return response;
+        });
+        lockService.unlock(response1.getToken());
+        latch.await();
+
+        forceSnapshot();
+
+        // T1 should hold nothing, T2 holds 3 in exclusive mode, T3 holds 1 in exclusive and 2 in shared mode
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1))
+                .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_3));
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_2))
+                .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_3));
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_3))
+                .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_2));
+
+        lockService.unlock(response2.getToken());
+        forceSnapshot();
+
+        // Only T3 should hold locks
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1))
+                .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_3));
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_2))
+                .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_3));
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_3)).isNull();
+
+        lockService.unlock(response3.get().getToken());
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isNull();
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_2)).isNull();
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_3)).isNull();
+    }
+
+    @Test
+    public void refreshDoesNotChangeSnapshot() throws InterruptedException {
+        LockResponse response =
+                lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, LOCK_1_THREAD_1_WRITE_REQUEST);
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isEqualTo(ANONYMOUS_TEST_THREAD_1);
+
+        lockService.refreshTokens(List.of(response.getToken()));
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isEqualTo(ANONYMOUS_TEST_THREAD_1);
+    }
+
+    @Test(timeout = 500L)
+    public void backgroundSnapshotRunnerWorks() throws InterruptedException {
+        LockServiceImpl lockServiceWithBackgroundRunner = LockServiceImpl.create(LockServerOptions.builder()
+                .isStandaloneServer(false)
+                .threadInfoConfiguration(ImmutableDebugThreadInfoConfiguration.builder()
+                        .recordThreadInfo(true)
+                        .threadInfoSnapshotIntervalMillis(10L)
+                        .build())
+                .build());
+        LockThreadInfoSnapshotManager backgroundSnapshotRunner = lockServiceWithBackgroundRunner.getSnapshotManager();
+
+        LockResponse response = lockServiceWithBackgroundRunner.lockWithFullLockResponse(
+                LockClient.ANONYMOUS, LOCK_1_THREAD_1_WRITE_REQUEST);
+
+        Thread.sleep(50L);
```

Comment:
Please use Awaitility! Alternatively, just create the SnapshotRunner yourself and passing in a DeterministicScheduler.